sycl/doc/extensions/ExtendedAtomics/SYCL_INTEL_extended_atomics.asciidoc

```diff
@@ -0,0 +1,525 @@
+
+== Dependencies
+
+This extension is written against the SYCL 1.2.1 specification, Revision v1.2.1-6.
+
+== Overview
+
+The SYCL atomic library (+cl::sycl::atomic+) defined in SYCL 1.2.1 is based on the standard atomic libary (+std::atomic+) but has some differences.  This extension introduces an alternative atomic class (+cl::sycl::intel::atomic+) including additional features from {cpp}11 and {cpp}20:
+
+- Overloaded operators to reduce the verbosity of using atomics
+- Support for floating-point types
+
+=== Overloaded Operators
+
+In SYCL 1.2.1, the +cl::sycl::atomic+ class provides atomic operations by way of member functions (e.g. +fetch_add+) without defining the corresponding operators (e.g. `+=`).  This increases the verbosity of simple uses of atomics, and requires developers to change their kernel code when switching between +read_write+ and +atomic+ accessors:
+
+[source,c++]
+----
+q.submit([&](handler& cgh)
+{
+  auto acc = buf.get_access<access::mode::atomic>(cgh);
+  cgh.parallel_for(..., [=](id<1> i)
+  {
+    acc[i].fetch_add(1); // developer cannot write acc[i]++ without this extension
+  });
+});
+----
+
+The operators defined by this extension match those defined for +std::atomic+ in {cpp}11.  The functionality of each operator is equivalent to calling a corresponding member function of the atomic class -- the operators do not expose any new functionality of the class, but act as shorthands for common use-cases.
+
+==== Operators for All Supported Types
+
+[source,c++]
+----
+  operator T();
+  T operator=(T desired);
+  T operator+=(T operand);
+  T operator-=(T operand);
+----
+
+==== Operators for Integral Types
+
+[source,c++]
+----
+  T operator++(int operand);
+  T operator--(int operand);
+  T operator++();
+  T operator--();
+  T operator&=(T operand);
+  T operator|=(T operand);
+  T operator^=(T operand);
+----
+
+=== Support for Floating-point Types
+
+In SYCL 1.2.1, support for floating-point types is limited to the +load+, +store+ and +exchange+ member functions.  Many applications requiring additional atomic operations (e.g. addition) currently work around this restriction using type punning and integer +compare_exchange+ operations.
+
+This extension extends support for floating-point types to the +compare_exchange+, +fetch_add+ and +fetch_sub+ functions in line with {cpp}20.  These new functions do not require dedicated floating-point atomic instructions and can be emulated using integer operations, giving compilers the freedom to choose the best implementation for the target device.
+
+== Modifications of SYCL 1.2.1 Specification
+
+=== Modify Sentence in Section 3.5.1
+
+==== From:
+
+Atomic access can also be requested on an accessor which allows +cl::sycl::atomic+ classes to be used via the accessor.
+
+==== To:
+
+Atomic access can also be requested on an accessor which allows +cl::sycl::intel::atomic+ classes to be used via the accessor.
```

Comment:
```suggestion
Atomic access can also be requested on an accessor which allows +cl::sycl::atomic+ and +cl::sycl::intel::atomic+ classes to be used via the accessor.
```

Or the idea of this extension is to prohibit usage of existing atomics from the SYCL standard? The same comment applies to several other places as well, where `cl::sycl::atomic` is replaced with `cl::sycl::intel::atomic`

Reply:
The idea was to replace `cl::sycl::atomic` with `cl::sycl::intel::atomic`, rather than have them both.  That way developers could use the new operators and member functions anywhere that they were previously using an atomic accessor, without having to go and change the accessor mode (e.g. to `intel::access::mode::atomic`, or something like that).  An accessor's `operator[]` can only return one thing, and it seems to make the most sense for it to be the new class here.

If I change "introduces" to "replaces" as suggested in your other comment, do you think that this would be clear?  Or do we need to expand on this elsewhere in the extension as well?

Reply:
> If I change "introduces" to "replaces" as suggested in your other comment, do you think that this would be clear?

This at least looks more uniform.

> The idea was to replace cl::sycl::atomic with cl::sycl::intel::atomic, rather than have them both. That way developers could use the new operators and member functions anywhere that they were previously using an atomic accessor, without having to go and change the accessor mode (e.g. to intel::access::mode::atomic, or something like that).

Got it, seems reasonable.
I'm still a bit confused with "extension", which "replaces" something instead of extending, but okay, let it be.

> An accessor's operator[] can only return one thing, and it seems to make the most sense for it to be the new class here.

I don't know if this should be mentioned in the spec or not, but we definitely need to have some switch in the implementation in order to be able to enable/disable this extension

Reply:
Thanks @AlexeySachkov.  I've made the introduces => replaces fix as suggested.

> I don't know if this should be mentioned in the spec or not, but we definitely need to have some switch in the implementation in order to be able to enable/disable this extension

I agree.  @bader, is there a conclusion from https://github.com/intel/llvm/issues/806?  Is the right thing to do here to add a note to this extension saying it's enabled by `-fsycl-extended-atomics` and can be tested using `__has_extension(sycl_extended_atomics)`?

Reply:
#806 is not finalized yet, suggested wording sounds reasonable.
I think we can say this in this document as this is just a proposal. I guess final version might have different wording once it's implemented and used.

Reply:
Great, thanks.  Added in 5e29012.